pkg/controller/job/job_controller.go

```diff
@@ -1094,12 +1103,27 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 //
 // Returns whether there are pending changes in the Job status that need to be
 // flushed in subsequent calls.
-func (jm *Controller) flushUncountedAndRemoveFinalizers(ctx context.Context, job *batch.Job, podsToRemoveFinalizer []*v1.Pod, uidsWithFinalizer sets.String, oldCounters *batch.JobStatus, podFailureCountByPolicyAction map[string]int, needsFlush bool) (*batch.Job, bool, error) {
+func (jm *Controller) flushUncountedAndRemoveFinalizers(ctx context.Context, job *batch.Job, podsToRemoveFinalizer []*v1.Pod, uidsWithFinalizer sets.String, oldCounters *batch.JobStatus, podFailureCountByPolicyAction map[string]int, needsFlush bool, newBackoffRecord backoffRecord) (*batch.Job, bool, error) {
 	var err error
 	if needsFlush {
 		if job, err = jm.updateStatusHandler(ctx, job); err != nil {
 			return job, needsFlush, fmt.Errorf("adding uncounted pods to status: %w", err)
 		}
+
+		key, _ := controller.KeyFunc(job)
+		err = jm.backoffRecordStore.updateBackoffRecord(backoffRecord{
+			key:                      key,
```

Comment:
it might be worth keeping track of what is currently in the uncounted set here.

Otherwise this is possible:

1. a pod fails
2. the pod is added to the uncounted set
3. the pod is accounted in failuresAfterLastSuccess
4. removing the finalizer fails
5. the next sync doesn't have the most up to date uncounted set (because this comes from the informer cache).
6. the failed pod is counted again.

Reply:
I think, if we have a stale entry in the sync, the update status will fail because of optimistic concurrency and the back off record doesn't get updated. Let me confirm this tomorrow day time and get back.

Reply:
ahh, that's a good point. That might be enough.